spec/type.dd

```diff
@@ -558,6 +558,38 @@ $(H3 $(LNAME2 string, $(D string)))
 
     $(P A $(DDSUBLINK spec/arrays, strings, $(I string) is a special case of an array.))
 
+$(H3 $(LNAME2 noreturn, $(D noreturn)))
+
+    $(P `noreturn` is the $(LINK2 https://en.wikipedia.org/wiki/Bottom_type, bottom type)
+    which can implicitly convert to any type, including `void`.
+    A value of type `noreturn` will never be produced and the compiler can
+    optimize such code accordingly.)
+
+    $(P A function that never returns has the return type `noreturn`. This can
+    occur due to an infinite loop or always throwing an exception.
+    This is particularly useful when inferring types.)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+noreturn abort(const(char)[] message);
+
+int example(int i)
+{
+    if (i < 0)
+    {
+        // abort does not return, so it doesn't need to produce an int
+        int val = abort("less than zero");
+    }
+    // ternary expression's common type is still int
+    return i != 0 ? 1024 / i : abort("calculation went awry.");
+}
+---
```

Comment:
This / another example should probably include usages of `noreturn` outside of function return types, i.e. variables.

Reply:
@MoonlightSentinel Do you have an idea for an example? The DIP only seems to mention `noreturn[] storage;` for `MapResult`, and stable dmd doesn't seem to have the enhanced compile-time understanding of this (which might make a good example).

Reply:
Unlike void, `noreturn` can be used everywhere a "normal" type may appear, i.e. variables, members, parameters, ... . 

Maybe add an example where the "return value" of a `noreturn` variable is stored in a variable?